File: Leetcode/cpp/Edit Distance.cc
class Solution {
public:
    static const int SIZE = 512;
    int dp[SIZE][SIZE];
    
    int minDistance(string word1, string word2) {
        memset(dp, 0 ,sizeof(dp));
        int len1 = word1.size();
        int len2 = word2.size();
        
        for (int i = 0; i <= len1; i++) {
            dp[i][0] = i;
        }
        
        for (int j = 0; j <= len2; j++) {
            dp[0][j] = j;
        }
        
        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                dp[i][j] = min(
                    min(
                        dp[i-1][j] + 1, 
                        dp[i][j-1] + 1
                    ),
                    dp[i-1][j-1] + (word1[i-1] == word2[j-1]? 0 : 1)
                );
            }
        }
        return dp[len1][len2];
    }
};

File: Leetcode/python/Edit Distance.py
class Solution:
    # @return an integer
    def minDistance(self, word1, word2):
        n, m = len(word1), len(word2)
        dp =[[0 for i in xrange(m + 1)] for j in xrange(n + 1)]
        
        for i in xrange(n + 1):
            dp[i][0] = i
        for i in xrange(m + 1):
            dp[0][i] = i
        
        for i in xrange(n):
            for j in xrange(m):
                delta = 1 if word1[i] != word2[j] else 0
                dp[i + 1][j + 1] = min(dp[i][j] + delta , dp[i][j + 1] + 1, dp[i + 1][j] + 1)
        return dp[n][m]
